### PR TITLE
cmd/zerosum: add zerosum tool

### PR DESCRIPTION
**Pull request by petermattis**
ZeroSum is a bank-like simulation that tests correctness in the face of
aggressive splits and lease transfers. Does not use `TestCluster` because I
want to eventually add node outages. Does not use docker because.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8034)

### CODE REVIEW

**Comment by petermattis** [comment]
Somewhat surprisingly, this hasn't found any problems after running for ~40m with the default settings. In that time, it created ~1000 ranges.

File: kv/db.go [hunk 1]
```diff
@@ -45,6 +45,8 @@ var allExternalMethods = [...]roachpb.Request{
 	roachpb.AdminSplit:       &roachpb.AdminSplitRequest{},
 	roachpb.AdminMerge:       &roachpb.AdminMergeRequest{},
 	roachpb.CheckConsistency: &roachpb.CheckConsistencyRequest{},
+	roachpb.RangeLookup:      &roachpb.RangeLookupRequest{},
+	roachpb.TransferLease:    &roachpb.TransferLeaseRequest{},
```
Comment (petermattis) [inline kv/db.go]: Is there a strong reason for disallowing these requests externally?

**Comment by tamird** [comment]
Rather than going this route, wouldn't it be possible to support node outages in TestCluster? We might even be able to support sophisticated network partitions via grpc's balancer abstraction.

**Comment by petermattis** [comment]
Can you sketch out how you would support a node outage in `TestCluster`? Stopping a per-node stopper is too graceful in my opinion.

**Comment by tschottdorf** [comment]
Reviewed 2 of 2 files at r1.
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[kv/db.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-KNbgdZFgpqaKIIrJpE6-r2-49:-KNbhkIvR2ZeMrGZ5cp_:2073878985) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc23c89f4146688deffa6ad7e5b92107886feee/kv/db.go#L49)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Is there a strong reason for disallowing these requests externally?
</details>
None for `RangeLookup` but strong reasons against `TransferLease`: any usage of it is illegal and instantly gives you a window for inconsistency. You must plumb your way to `(*Replica).AdminTransferLease`, i.e. introduce a new admin command (similar to `AdminSplit`), and then expose that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034)*

**Comment by tschottdorf** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[cmd/zerosum/cluster.go, line 222 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-KNbjGb5a6DErSCvPzWv:-KNbjGb6wpqvMnO9kD4R:-2048261480) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc23c89f4146688deffa6ad7e5b92107886feee/cmd/zerosum/cluster.go#L222)):*
> ```Go
> 
> 	// TODO(peter): This is using too much knowledge about the internals of
> 	// leases.
> ```

This should be handled if you use the correct entry point `(*Replica).AdminTransferLease`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[kv/db.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-KNbgdZFgpqaKIIrJpE6-r2-49:-KNbtkmIaDJ9W5IGtVPv:1916076563) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc23c89f4146688deffa6ad7e5b92107886feee/kv/db.go#L49)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> None for `RangeLookup` but strong reasons against `TransferLease`: any usage of it is illegal and instantly gives you a window for inconsistency. You must plumb your way to `(*Replica).AdminTransferLease`, i.e. introduce a new admin command (similar to `AdminSplit`), and then expose that.
</details>
Eh? Doesn't `TransferLease` perform all the appropriate checks that the lease transfer is valid? If not, what prevents a delayed `TransferLease` request from causing havoc?

Plumbing in support for an `AdminTransferLease` makes sense, though I'd like to better understand the danger of `TransferLease`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034)*

**Comment by petermattis** [comment]
Another caveat with `TestCluster` is that it creates in-memory nodes. We could plumb through support for disk-based nodes, just more work.

**Comment by tschottdorf** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[kv/db.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-KNbgdZFgpqaKIIrJpE6-r2-49:-KNbz1xaCKLBWexn3v9X:-346196648) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc23c89f4146688deffa6ad7e5b92107886feee/kv/db.go#L49)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Eh? Doesn't `TransferLease` perform all the appropriate checks that the lease transfer is valid? If not, what prevents a delayed `TransferLease` request from causing havoc?
> 
> Plumbing in support for an `AdminTransferLease` makes sense, though I'd like to better understand the danger of `TransferLease`. 
</details>
No, `TransferLease` is proposed exclusively by the replica which transfers its lease away, where it promises at the time of proposal not to use that lease any more. No such promise is associated with what you're doing, so the old lease holder will happily serve reads which run contrary to the timestamp cache.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[kv/db.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-KNbgdZFgpqaKIIrJpE6-r2-49:-KNc59mcsX0It9mADHCg:1146162818) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc23c89f4146688deffa6ad7e5b92107886feee/kv/db.go#L49)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> No, `TransferLease` is proposed exclusively by the replica which transfers its lease away, where it promises at the time of proposal not to use that lease any more. No such promise is associated with what you're doing, so the old lease holder will happily serve reads which run contrary to the timestamp cache.
</details>
Do you have any explanation for why I'm not seeing problems in my test? I verified (via logs) that lease transfers are happening.

Looks to me like a `TransferLeaseRequest` received externally will be redirected to the range lease holder (because `TransferLease` is a "write" command). I'm not terribly familiar with this code so I'm undoubtedly missing something. Please enlighten me, Sensei. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034)*

**Comment by tschottdorf** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[kv/db.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-KNbgdZFgpqaKIIrJpE6-r2-49:-KNcAhnYT0q14sRiUEVq:957479349) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc23c89f4146688deffa6ad7e5b92107886feee/kv/db.go#L49)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Do you have any explanation for why I'm not seeing problems in my test? I verified (via logs) that lease transfers are happening.
> 
> Looks to me like a `TransferLeaseRequest` received externally will be redirected to the range lease holder (because `TransferLease` is a "write" command). I'm not terribly familiar with this code so I'm undoubtedly missing something. Please enlighten me, Sensei. 
</details>
That can only be due to the workload not tickling it. `TransferLease` will execute on all Replicas alike, but as stated above, the magic in `TransferLease` is that the lease holder gives a promise when it *proposes* it. It will still propose it with what you are doing (at least in the majority of cases), but the promise isn't given.

Essentially the lease must be created in `r.mu.pendingLeaseRequest.InitOrJoinRequest`, which makes the Replica aware that there's a pending lease (and potential lease transfer).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034)*

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034)*

**Comment by petermattis** [comment]
Review status: 1 of 41 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[cmd/zerosum/cluster.go, line 222 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-KNbjGb5a6DErSCvPzWv:-KNe2XQZ1CIuWYuq4y-8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc23c89f4146688deffa6ad7e5b92107886feee/cmd/zerosum/cluster.go#L222)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This should be handled if you use the correct entry point `(*Replica).AdminTransferLease`.
</details>
Done.

---

*[kv/db.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-KNbgdZFgpqaKIIrJpE6-r2-49:-KNe2_rLoKtomKi3E94Q:-1443091937) ([raw file](https://github.com/cockroachdb/cockroach/blob/abc23c89f4146688deffa6ad7e5b92107886feee/kv/db.go#L49)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> That can only be due to the workload not tickling it. `TransferLease` will execute on all Replicas alike, but as stated above, the magic in `TransferLease` is that the lease holder gives a promise when it *proposes* it. It will still propose it with what you are doing (at least in the majority of cases), but the promise isn't given.
> 
> Essentially the lease must be created in `r.mu.pendingLeaseRequest.InitOrJoinRequest`, which makes the Replica aware that there's a pending lease (and potential lease transfer).
</details>
Ok, I am now enlightened. See #8050 where I make `AdminTransferLease` a part of `BatchRequest` so that it can be called via RPC. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034)*

**Comment by petermattis** [comment]
PS The first commit is #8050 and can be ignored in this PR. 

---

Review status: 1 of 41 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-:-KNe2ypg--rgJBO_ADrx:679275391)*

**Comment by petermattis** [comment]
This is ready for another look. I'm not sure what is going on with reviewable, but only the files in `cmd/zerosum` remain present in the PR.

I'd like some opinions on whether this should live in `cmd/` or somewhere else. I could make this an acceptance test, just one that doesn't use docker. 

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-:-KNiEUDpH9Mv8qZqiiEj:107220144)*

**Comment by petermattis** [comment]
I've added another commit which introduces a chaos agent. Currently this agent periodically kills the first node, leaves it dead for 15-45s and then restarts it. This causes various bits of bad performance (zero ops/sec for up to a minute) and also produced on inconsistency where the total balance in the accounts became `-1`. Seems like a worthwhile tool to be running frequently. Let's figure out the right place for it to live.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-:-KNmZeRBjAu0Rs3u9FXK:2016278086)*

**Comment by bdarnell** [comment]
:lgtm:

Note that @mrtracy just added the ability to stop individual servers in `TestCluster`, which might remove the need for this new multi-process abstraction.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-:-KNtcQFyrRsju7WN-xly:-1456689465)*

**Comment by petermattis** [comment]
Yes, `TestCluster` can now stop a server, but that server can't be restarted. The servers run by `TestCluster` are in memory so state is lost when they are stopped. This could be fixed and perhaps should be fixed, but as it stands `TestCluster` couldn't be used here.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8034#-:-KNw8UQpP4JzwZbOkKom:146986247)*

**Comment by bdarnell** [comment]
Sigh. We should probably look at moving `storage.multiTestCluster` up the stack instead of reimplementing its functionality piecemeal in `TestCluster`.

**Pull request merged**